trestle/core/control_io.py

```diff
@@ -985,88 +1041,162 @@ def _comp_name_in_dict(comp_name: str, comp_dict: Dict[str, List[Dict[str, str]]
+                        'Please combine the sections so there is only one heading for each component in a statement.'
+                    )
+                comp_list.append(simp_comp_name)
+                comp_name = ControlIOReader._comp_name_in_dict(comp_name, comp_dict)
+            if comp_name in comp_dict:
+                if label in comp_dict[comp_name]:
+                    comp_dict[comp_name][label].prose += '\n' + prose
+                else:
+                    comp_dict[comp_name][label] = ComponentImpInfo(prose=prose)
             else:
-                comp_dict[comp_name][label] = prose
-        else:
-            comp_dict[comp_name] = {label: prose}
+                comp_dict[comp_name] = {label: ComponentImpInfo(prose=prose)}
+            # keep track of subnodes already handled
+            subnode_kill: List[int] = []
+            for ii, subnode in enumerate(node.subnodes):
+                if subnode.key.find(const.IMPLEMENTATION_STATUS_HEADER) >= 0:
+                    # ### Implementation Status: implemented
+                    status = subnode.key.split(maxsplit=3)[-1]
+                    comp_dict[comp_name][label].implementation_status = ControlIOReader._imp_stat_from_string(status)
+                    subnode_kill.append(ii)
+                elif subnode.key.split(maxsplit=2)[1] == const.REMARKS:
+                    comp_dict[comp_name][label].remarks = subnode.content.text
+                    subnode_kill.append(ii)
+            for index in subnode_kill:
+                del node.subnodes[index]
         for subnode in node.subnodes:
-            ControlIOReader._add_node_to_dict(comp_name, label, comp_dict, subnode, control_id, comp_list)
+            ControlIOReader._add_node_to_dict(
+                comp_name, label, comp_dict, subnode, control_id, comp_list, component_mode
+            )
+
+    @staticmethod
+    def _imp_stat_from_string(stat_str: str) -> str:
+        """Find best matching status based on string."""
+        simp_str = ControlIOReader.simplify_name(stat_str)
+        for stat_const in const.STATUS_ALL:
+            if simp_str == ControlIOReader.simplify_name(stat_const):
+                return stat_const
+        logger.warning(f'Unrecognized status marked as unknown: {stat_str}')
+        return const.STATUS_TRESTLE_UNKNOWN
+
+    @staticmethod
+    def _add_component_to_dict(
+        control_id: str, comp_dict: CompDict, comp_def: Optional[comp.ComponentDefinition], comp_name: Optional[str]
+    ) -> None:
+        """Strictly works with ComponentDefinition elements and not SSP."""
+        if comp_def:
+            for sub_comp in as_list(comp_def.components):
+                if sub_comp.title == comp_name:
+                    for control_imp in as_list(sub_comp.control_implementations):
+                        for imp_req in as_list(control_imp.implemented_requirements):
+                            if imp_req.control_id == control_id:
```

Comment:
Please refactor this...

Reply:
This has been split into separate lower level function calls.